Rstudio 2023 09 0 463

Kupac opened the pull request:
## Update and fix RStudio

- Fixed plotting issue
- Added quarto support
- Upgraded to current version (2023.09.0+463 "Desert Sunflower")

Fixes https://github.com/NixOS/nixpkgs/issues/237878 https://github.com/NixOS/nixpkgs/issues/194071, maybe more

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

nagy commented:
Hi, do you want to put this PR in draft mode while you are figuring things out? eventually it should result in one commit if you are updating one package.

Kupac commented:
Good point, I messed it up big time

Kupac commented on pkgs/development/r-modules/wrapper-rstudio.nix:
```diff
@@ -2,54 +2,69 @@
+        '')
+    );
+in
+symlinkJoin {
```
Quarto needs to be in the user's path, otherwise RStudio gives a warning. Since quarto is  a dep, and it's in the nix store anyway, it's not costly to link it in the path

Kupac replied:
Tagging @nviets @b-rodrigues

nviets commented:
Result of `nixpkgs-review pr 252047` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>rstudio</li>
    <li>rstudio-server</li>
    <li>rstudioServerWrapper</li>
    <li>rstudioWrapper</li>
  </ul>
</details>

jbedo commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -26,6 +27,7 @@
 , pandoc
 , llvmPackages
 , yaml-cpp
+, yarn
```
```suggestion
```
Yarn is patched out.

jbedo commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -89,6 +102,7 @@ in
       makeWrapper
       pandoc
       nodejs
+      yarn
```
```suggestion
```

nviets commented:
We need to confirm that quarto knitting and visual mode work in this build - both server and application modes. Neither worked for me when I tried this branch. I'll give it another try soon. Thanks, @Kupac for taking this further.

RStudio only supports R 4.3 as of [2023.03+ versions](https://posit.co/blog/rstudio-2023-03-0-whats-new/), so this upgrade really needs some attention.

jbedo commented:
It does not work for me currently:

> Error: pandoc version 1.12.3 or higher is required and was not found (see the help page ?rmarkdown::pandoc_available).

Kupac commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -61,17 +67,24 @@ let
     inherit pname version;
     packageJSON = ./package.json;
-    yarnLock = "${src}/src/gwt/panmirror/src/editor/yarn.lock";
+    yarnLock = ./yarn.lock;
```
Upstream deleted the panmirror folder, and put yarn.lock in gitignore. I think they manage that file locally, and it disappeared form the git repo. Someone (@nviets ?) placed it in the nixpkgs repo instead.

nviets replied:
Yeah, that was me. I was fumbling through this yarn part, so you're welcome to do whatever works. There is some discussion on that here: https://github.com/NixOS/nixpkgs/pull/245269.

jbedo replied:
Where did the lock file come from then? This should point to that location to avoid bundling this into nixpkgs.

Kupac replied:
may be simply a copy of the lock file from before? @nviets do you remember how you got it?

Kupac replied:
There's actually a node.js-related error, although it doesn't seem to be a problem. The OCD in me would still like it to go away. If anybody has ideas how to fix it, please don't hold back.

     [exec] Error: Cannot find module '/nix/store/gv66dg8gjckw42m0iwnyh9zx440da9bl-source/apps/panmirror/fuse'
     [exec]     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
     [exec]     at Module._load (node:internal/modules/cjs/loader:922:27)
     [exec]     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
     [exec]     at node:internal/main/run_main_module:23:47 {
     [exec]   code: 'MODULE_NOT_FOUND',
     [exec]   requireStack: []
     [exec] }
     [exec] 
     [exec] Node.js v18.17.1
     [exec] Result: 1

nviets replied:
See my original PR for how to get the lock file. Someone explains there

jbedo replied:
I think this whole panmirror bit is broken. The modules is now part of the `quarto` source and in principle the following should work:

```
  panmirrorModules = mkYarnModules rec {
    pname = "panmirror";
    version = "dev";
    packageJSON = "${quartoSrc}/apps/panmirror/package.json";
    yarnLock = "${quartoSrc}/yarn.lock";
    offlineCache = fetchYarnDeps {
      inherit yarnLock;
      hash = "sha256-dykyuVoMyz0aTZkXmspfPL9yMPGJQYGGZ+Y8MMM3cpQ=";
    };
  };
```

It doesn't though, fails because it can't find the `editor` module which is also part of the quarto source. I don't know enough about the yarn/node to progress much further at the moment.

Kupac commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -40,17 +42,21 @@
-  RSTUDIO_VERSION_MINOR  = "07";
-  RSTUDIO_VERSION_PATCH  = "1";
-  RSTUDIO_VERSION_SUFFIX = "+554";
+  version =
+          RSTUDIO_VERSION_MAJOR +
+    "." + RSTUDIO_VERSION_MINOR +
+    "." + RSTUDIO_VERSION_PATCH +
+          RSTUDIO_VERSION_SUFFIX;
```
I'll fix this. Just to learn something: what's wrong with string concat with +?

jbedo replied:
Nothing as such, it's functionally equivalent after all. The interpolated version is easier to read though.

Kupac commented:
> We need to confirm that quarto knitting and visual mode work in this build - both server and application modes. Neither worked for me when I tried this branch. I'll give it another try soon. Thanks, @Kupac for taking this further.

Interesting. Knitting worked for me, but will try again. Visual mode never worked, and we agreed on Matrix that rstudio is already useful without it, so we should release without. Unless you have some ideas how to debug. I only see a spinning loading wheel.

nviets commented:
No idea how to debug visual mode. Ideally we need to get the attention of someone on Posit's IDE team. Posit Conference is next week in Chicago, in case anyone is in the area.

Kupac commented:
> It does not work for me currently:
> 
> > Error: pandoc version 1.12.3 or higher is required and was not found (see the help page ?rmarkdown::pandoc_available).

I figured out that pandoc is detected via the `RSTUDIO_PANDOC` envvar. But it should not point to `${pandoc}/bin/pandoc`, but to `${pandoc}/bin`. I'm trying to change it in the sessionConstants.hpp, let's see if that works. Maybe the same must be done with quarto?

Kupac commented:
OK, that worked. I think I'll try the same with quarto, maybe the warning goes away? (I don't think so)

jbedo commented:
Note that pandoc linking with `rPackages.rmarkdown` should at least now be fixed: #255219.

nviets commented:
Thanks, @jbedo, for the markdown fix. I will rebase and give it a try by tomorrow.

nviets commented:
Had a go with this build in the application mode (not seerver). I am able to use the Render button with a qmd file now, which is great. Probably good enough for now. In the Render log, it appears to be using pandoc, and I'm not sure if that's to be expected, since quarto depends on pandoc under the hood.

I still see the Quarto CLI warning at the bottom of the screen. @Kupac - I thought you said this warning disappeared when you included quarto in the user bin, no?

It's late where I am, and I'll have a closer look tomorrow after some sleep.

Kupac commented:
So here's my trace back on the quarto error:

It comes from [SessionQuarto.cpp](https://github.com/rstudio/rstudio/blob/1fb8c970fb212301ce2e91834f2b68deb7827169/src/cpp/session/modules/quarto/SessionQuarto.cpp#L125). Quarto version is actually detected later on in the function [readQuartoVersion](https://github.com/rstudio/rstudio/blob/1fb8c970fb212301ce2e91834f2b68deb7827169/src/cpp/session/modules/quarto/SessionQuarto.cpp#L94) by starting from the quarto binary path, going up 2 levels, and reading the share/version file. This is all in the nix derivation: so far so good.

What we have here, is a user installed quarto (i.e. not bundled with rstudio). It can be detected in one of two ways.

  1. [FilePath](https://github.com/rstudio/rstudio/blob/1fb8c970fb212301ce2e91834f2b68deb7827169/src/cpp/session/modules/quarto/SessionQuarto.cpp#L141C42-L141C53), which looks for quarto in ${PATH}. This is why symlinkJoin worked, because quarto was in the path
  2. [RSTUDIO_QUARTO](https://github.com/rstudio/rstudio/blob/1fb8c970fb212301ce2e91834f2b68deb7827169/src/cpp/session/modules/quarto/SessionQuarto.cpp#L143) environment variable. If I start RStudio from this branch with setting RSTUDIO_QUARTO to the quarto binary path, the warning disappears.

I think option 2 would be preferred, right? What would be the best way to achieve that? Is it OK to set this env var?  @nviets or @jbedo ?

nviets commented:
Great detective work! Based on [the logic](https://github.com/rstudio/rstudio/blob/1fb8c970fb212301ce2e91834f2b68deb7827169/src/cpp/session/modules/quarto/SessionQuarto.cpp#L154) you found in the source, I think Option 2 makes the most sense.

Kupac commented:
OK, I made a new commit based on what we discussed yesterday. I used wrapProgram to set the env var to the quarto path, and the warning is gone. I hope that's acceptable?

nviets commented:
Looks good to me. I think we have more to do with RStudio, especially fixing the mac build and visual mode, but these would be better handled in separate PRs since they are already broken in the current build. No need to delay this valuable update. Great work @Kupac - thanks for taking this across the finish line!

jbedo commented:
It's not quite right yet, but I haven't figured it out. I don't think the wrapper should be needed if the session constants work properly. I've been working on it today but it's slow going as builds take a while.

Kupac commented:
Exactly! The pandoc part is using a session constant, which somehow sets
the rstudio_pandoc env var, which can then be overridden by the user. But I
couldn't figure out how to do the same for quarto. Code looks similar, but
env var doesn't exist. I agree that it would be more appropriate that way.

Justin Bedő ***@***.***> ezt írta (időpont: 2023. szept. 18.,
Hét 5:24):

> It's not quite right yet, but I haven't figured it out. I don't think the
> wrapper should be needed if the session constants work properly. I've been
> working on it today but it's slow going as builds take a while.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/252047#issuecomment-1722703765>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6X6OK3POL234F2HBAMER3X265HNANCNFSM6AAAAAA4CA3TXM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jbedo commented on pkgs/applications/editors/rstudio/do-not-install-pandoc.patch:
```diff
@@ -1,13 +1,46 @@
+ # - if quarto is not enabled, use pandoc fallback
+ if(QUARTO_ENABLED)
+-   set(RSTUDIO_DEPENDENCIES_PANDOC_DIR "${RSTUDIO_DEPENDENCIES_QUARTO_DIR}/bin/tools")
++   set(RSTUDIO_DEPENDENCIES_PANDOC_DIR "@pandoc@/bin/pandoc")
```
```suggestion
+   set(RSTUDIO_DEPENDENCIES_PANDOC_DIR "@pandoc@/bin")
```

jbedo commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -206,14 +218,19 @@ in
     '';
 
-    meta = with lib; {
+    postFixup = ''
+      wrapProgram $out/bin/rstudio \
+        --set RSTUDIO_QUARTO ${quarto}/bin/quarto
+    '';
```
```suggestion
```
Not needed with suggestion above.

jbedo reviewed (changes requested):
Couple more things:

1. The changes to rstudioWrapper are formatting, please revert this file.
2. The file names for the patches no-longer reflect their function. Can you rename them more accurately?

On the panmirror issues: I'm inclined to strip all this out and accept visual mode will be broken for the moment. What do you think?

b-rodrigues commented:
>On the panmirror issues: I'm inclined to strip all this out and accept visual mode will be broken for the moment. What do you think?

Personally, I would be ok with this. I think that not having visual mode will likely affect no one for the time being. We’ll see how we can deal with this later, or maybe try to get some help from Posit as @nviets suggested.

nviets commented:
Unless someone gets to it first, I'll open an issue on rstudio/rstudio to ask for a little help with Visual Mode. Hopefully we'll get some traction from the developer team. Can't hurt to ask.

jbedo commented:
> Unless someone gets to it first, I'll open an issue later today on rstudio/rstudio to ask for a little help with Visual Mode. Hopefully we'll get some traction from the developer team. Can't hurt to ask.

This is not an upstream issue, we need knowledge of how node packages are managed in nixpkgs. We probably to run yarn manually rather than using `mkYarnModules`.

nviets commented:
@jbedo - are you saying that visual mode is failing because of something with yarn? I was under the impression we needed more patching in the source files. From the glance I had, it looks like visual mode [needs to manage state files](https://github.com/rstudio/rstudio/blob/main/src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/visualmode/VisualModeEditingLocation.java#L38), and I wonder if that mechanic is working correctly in the nix'd build. I also see panmirror references in the same file.

jbedo commented:
There might be more things wrong with visual mode but panmirror is
completely broken currently. Fixing it requires some knowledge about
node, yarn and how these are handled in nixpkgs.

Kupac commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -144,16 +151,20 @@ in
-      sed '1i#include <set>' -i src/cpp/core/include/core/Thread.hpp
+      substituteInPlace src/cpp/session/include/session/SessionConstants.hpp \
+        --replace '@pandoc@' ${pandoc}/bin \
+        --replace '@quarto@' ${quarto}/bin
```
Weird how it needs to be pandoc/bin but quarto without /bin. Not too consistent, but as long as it works...

Commit:
rstudio: 2022.07.1+554 -> 2023.09.0+463

Updates rstudio to current version, fixes plotting and enables quarto
and markdown rendering.
Visual mode for markdown editing is still not functional.

 * Remove unneeded patch
 * Eliminate false quarto version warning
 * Fix pandoc and quarto paths
 * Adjust patch name to reflect purpose
 * Remove panmirror for now
 * Update quartoSrc

Co-authored-by: Justin Bedő <cu@cua0.org>
Also contributed: b-rodrigues and kupac

Kupac commented:
> On the panmirror issues: I'm inclined to strip all this out and accept visual mode will be broken for the moment. What do you think?
Absolutely. If it compiles, I'm all for it. But how does one do this?

Kupac commented:
>     2. The file names for the patches no-longer reflect their function. Can you rename them more accurately?

Can do, but which ones do you mean? `do-not-install-pandoc` looks like a complete misnomer to me, but the others are OK IMO.

jbedo commented:
> > On the panmirror issues: I'm inclined to strip all this out and accept visual mode will be broken for the moment. What do you think?
> 
> Absolutely. If it compiles, I'm all for it. But how does one do this?

Just remove the `panModules` bit, and where it copies the output to the source tree. It's not currently copying to the right location anyway so it shouldn't affect the build.

> > ```
> > 2. The file names for the patches no-longer reflect their function. Can you rename them more accurately?
> > ```
> 
> Can do, but which ones do you mean? `do-not-install-pandoc` looks like a complete misnomer to me, but the others are OK IMO.

Yes sorry, I mean `do-not-install-pandoc` should be renamed. Maybe `use-system-quarto`?

Kupac commented:
OK, should be all done. @jbedo you were right, removing all the panmirror things didn't affect anything. That's good, because it compiled much faster, hehe. So now we're ready. At least I am. will keep rebasing until someone merges this

jbedo commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -61,18 +62,15 @@ let
+  quartoSrc = fetchFromGitHub {
+    owner = "quarto-dev";
+    repo = "quarto";
+    rev = "a92906bbd004e3d8abc39b5f59346ca2a091d1e8";
```
Can you double check this? I think this version of rstudio is mountain hydrangea, and the `rstudio-mountain-hydrangea` branch of `quarto` is currently [01345470a8f80becb1e128be24f5](https://github.com/quarto-dev/quarto/commit/01345470a8f80becb1e128be24f59d2c34fb3a85).

Kupac replied:
I think it's "Elsbeth Geranium" , but I got the idea :)

Kupac replied:
Because there are only 2 release branches, and the later one is mountain-hydrangea. They don't have one for "Elsbeth Geranium"

Kupac replied:
Actually, the hydrangea release was _earlier_ than the current rev in the default.nix. So I looked a bit more into it and chose a rev that is close to the release date of rstudio. Probably it's not exactly what Posit is using, but close enough. Everything still works, except visual mode.

jbedo commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -243,3 +243,4 @@ in
       })
     ];
   })
+
```
```suggestion
```

jbedo commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -206,14 +205,15 @@ in
     '';
 
-    meta = with lib; {
+
```
```suggestion
```

jbedo reviewed (commented):
Just that hash to check, otherwise LGTM.

jbedo commented on pkgs/applications/editors/rstudio/default.nix:
```diff
@@ -40,39 +40,39 @@
-      inherit yarnLock;
-      hash = "sha256-v05Up6VMlYlvgUYQVYo+YfpcsMohliNfMgyjq6QymCI=";
-    };
+  # rev should be something close to the rstudio release date
```
Last nit: is this what we should be aiming for? Is it more appropriate to use the corresponding release branches? I understand this is how this revision was chosen, just trying to understand what it should be for future updates.

Kupac replied:
There are only 2 `release/*` branches, none of which is for the current rstudio nickname. so they don't do this consistently for each release. I thought the best we can do is to take the quarto commit with a version bump that precedes the rstudio release - and test that it works. That's how I chose the rev. What do you think? Should I improve the comment?

jbedo replied:
In light of that I think choosing closest revision to release is fine.

Kupac commented:
No waaaaay https://github.com/rstudio/rstudio/releases/tag/v2023.09.0%2B463

Kupac commented:
Time to update the PR ;-)
![image](https://github.com/NixOS/nixpkgs/assets/8224569/23272f9e-1eb8-461f-8b42-6bfbd5f3d3c6)

jbedo commented:
> Edit: I don't seem to be able to change the source branch. But it's here, and it works. https://github.com/Kupac/nixpkgs/tree/rstudio_2023_09_0_463 Somebody can advise what to do?

You can reset this branch to the updated commit (`git reset --hard rstudio_2023_09_0_463`) and force push to update this PR.

hervyqa commented:
maybe fix #210826

Kupac commented:
> maybe fix #210826

Nope, not yet. This one is quite hard to debug, bcz there's no error message. We may need help from posit developers. You know anyone?
